# Métodos

## Métodos de conversão

In [ ]:
import numpy as np

In [ ]:
array_3d = np.arange(1, 25)
array_3d

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])

In [ ]:
array_3d = np.arange(1, 25).reshape(2, 4, 3) # profundidade 2, linhas 4, colunas 3
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

### `tolist`

Transforma o array para o formato de lista

In [ ]:
array_3d.tolist()

[[[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]],
 [[13, 14, 15], [16, 17, 18], [19, 20, 21], [22, 23, 24]]]

In [ ]:
type(array_3d.tolist())

list

### `item`

Retorna um elemento específico do array.

In [ ]:
array_3d[1, 0, 2]

15

In [ ]:
array_3d.item((1, 0, 2)) # passando a localização do item (profundidade, linha, coluna)

15

In [ ]:
array_3d.item(14) # passando o índice

15

### `itemset`

Insere um valor escalar em um array

In [ ]:
# substituindo o item da posição 14 por outro valor
array_3d.itemset(14, 100) # troca o valor do item 14 (troca 15 por 100)
array_3d

array([[[  1,   2,   3],
        [  4,   5,   6],
        [  7,   8,   9],
        [ 10,  11,  12]],

       [[ 13,  14, 100],
        [ 16,  17,  18],
        [ 19,  20,  21],
        [ 22,  23,  24]]])

In [ ]:
# passa as coordenadas para troca o valor do item 14 (troca 15 por 99)
array_3d.itemset((1, 0, 2), 99)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 99],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

### `view`

Cria uma cópia rasa do array, ou seja, quando a cópia do array é modificada, o array original também é modificado.

In [ ]:
# recriando o array inicial
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_view = array_3d.view()
array_view

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_view[0, 0, 0] = 99 # modificando o primeiro elemento
array_view

array([[[99,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
# Observe que o original também foi modificado
array_3d

array([[[99,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

### `copy`

Cria uma cópia profunda, ou seja, modificar os valores da cópia, não modifica o array original.

In [ ]:
# recriando o array inicial
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_copy = array_3d.copy()
array_copy

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_copy[0, 0, 0] = 99 # modificando o primeiro elemento
array_copy

array([[[99,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
# Observe que o original não foi modificado
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

### `fill`

Preenche o array com um valor escalar

In [ ]:
# recriando o array inicial
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_3d.fill(99)
array_3d

array([[[99, 99, 99],
        [99, 99, 99],
        [99, 99, 99],
        [99, 99, 99]],

       [[99, 99, 99],
        [99, 99, 99],
        [99, 99, 99],
        [99, 99, 99]]])

A documentação completa dos métodos de conversão: https://numpy.org/doc/stable/reference/arrays.ndarray.html#array-conversion

## Métodos de manipulação de shape

### `reshape`

Muda as dimensões do array. O novo formato deve ser compatível com a quantidade de elementos do array.

In [ ]:
# recriando o array 3d
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_3d.shape

(2, 4, 3)

In [ ]:
array_3d.reshape(4, 3, 2)

array([[[ 1,  2],
        [ 3,  4],
        [ 5,  6]],

       [[ 7,  8],
        [ 9, 10],
        [11, 12]],

       [[13, 14],
        [15, 16],
        [17, 18]],

       [[19, 20],
        [21, 22],
        [23, 24]]])

In [ ]:
# array_3d.reshape(5, 2, 1)
# ValueError: cannot reshape array of size 24 into shape (5,2,1)
# 5 x 2 x 1 é diferente de 24

In [ ]:
# -1 faz o numpy inferir o tamanho do eixo com base nos demais eixos
# como há 4 profundidade, 3 linhas o numpy infere que o numero de colunas é 2
# pois assim o total é 24.

array_3d.reshape(4, 3, -1)

array([[[ 1,  2],
        [ 3,  4],
        [ 5,  6]],

       [[ 7,  8],
        [ 9, 10],
        [11, 12]],

       [[13, 14],
        [15, 16],
        [17, 18]],

       [[19, 20],
        [21, 22],
        [23, 24]]])

In [ ]:
# -1 faz o numpy inferir o tamanho do eixo com base nos demais eixos
# como há 4 profundidade, 2 colunas o numpy infere que o numero de linhas é 3
# pois assim o total é 24.

array_3d.reshape(4, -1, 2)

array([[[ 1,  2],
        [ 3,  4],
        [ 5,  6]],

       [[ 7,  8],
        [ 9, 10],
        [11, 12]],

       [[13, 14],
        [15, 16],
        [17, 18]],

       [[19, 20],
        [21, 22],
        [23, 24]]])

### `transpose`

Faz a operação matricial de transposição

In [ ]:
# visualizando a matriz original
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_3d.shape

(2, 4, 3)

In [ ]:
# 1: indice do array_3d.shape que será a dimensão de produndidade, nesse caso 4
# 0: indice do array_3d.shape que será a dimensão de linhas, nesse caso 2
# 2: indice do array_3d.shape que será a dimensão de coluna, nesse caso 3

# indice:   0  1  2
#          (2, 4, 3)
array_transpose = array_3d.transpose((1, 0, 2)) # 4 profundidade, 2 linhas, 3 colunas
array_transpose

array([[[ 1,  2,  3],
        [13, 14, 15]],

       [[ 4,  5,  6],
        [16, 17, 18]],

       [[ 7,  8,  9],
        [19, 20, 21]],

       [[10, 11, 12],
        [22, 23, 24]]])

In [ ]:
# criando um array 2d para entender melhor
array_2d = np.array([[1, 2, 3], [4, 5, 6]])
array_2d

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
array_2d.transpose((1, 0)) # faz uma rotação

array([[1, 4],
       [2, 5],
       [3, 6]])

In [ ]:
array_2d.reshape(3, 2) # faz uma recriação mantendo a ordem dos dados

array([[1, 2],
       [3, 4],
       [5, 6]])

### `swapaxes`

Muda dois eixos de lugar

In [ ]:
# recriando o array original
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
# O eixo das linhas (1) será transformado no eixo da profundidade
# O eixo da profundidade (0) será tranformado no eixo das linhas
array_3d.swapaxes(1, 0)

array([[[ 1,  2,  3],
        [13, 14, 15]],

       [[ 4,  5,  6],
        [16, 17, 18]],

       [[ 7,  8,  9],
        [19, 20, 21]],

       [[10, 11, 12],
        [22, 23, 24]]])

In [ ]:
# tem o mesmo efeito do seguinte transpose
array_3d.transpose((1, 0, 2))

array([[[ 1,  2,  3],
        [13, 14, 15]],

       [[ 4,  5,  6],
        [16, 17, 18]],

       [[ 7,  8,  9],
        [19, 20, 21]],

       [[10, 11, 12],
        [22, 23, 24]]])

### `flatten`

Retorna uma cópia do array em um única dimensão.

In [ ]:
array_3d.flatten() # retorna um copia em uma nova variavel (não interfere no original)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])

### `ravel`

Retorna uma view do array em um única dimensão.

In [ ]:
array_3d.ravel() # retorna um view (mudar os valores interfere no original)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])

In [ ]:
# é possível ter o mesmo resultado com reshape
array_3d.reshape(-1)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])

### `squeeze`

Remove os eixos de tamanho 1

In [ ]:
array_axis_size_1 = array_3d.reshape(8, 1, 3) # profundidade 8, linhas 1, colunas 3
array_axis_size_1

array([[[ 1,  2,  3]],

       [[ 4,  5,  6]],

       [[ 7,  8,  9]],

       [[10, 11, 12]],

       [[13, 14, 15]],

       [[16, 17, 18]],

       [[19, 20, 21]],

       [[22, 23, 24]]])

In [ ]:
array_axis_size_1.shape

(8, 1, 3)

In [ ]:
# cria uma matriz com 8 linhas e 3 colunas
# removendo a dimensão de valor 1 (linhas)
array_squeeze = array_axis_size_1.squeeze()
array_squeeze

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12],
       [13, 14, 15],
       [16, 17, 18],
       [19, 20, 21],
       [22, 23, 24]])

Documentação dos métodos de manipulação de `shape`: https://numpy.org/doc/stable/reference/arrays.ndarray.html#shape-manipulation

## Métodos de seleção e manipulação de itens

### `take`

Retorna os valores dos indices solicitados

In [ ]:
# recriando o array_3d
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
# indice 1 = valor 2
# indice 5 = valor 6
# indice 10 = valor 11
# indice 12 = valor 13
array_3d.take(indices=[1, 5, 10, 12])

array([ 2,  6, 11, 13])

In [ ]:
# apresentando o resultado em outro no formato
# pasta específicar no paramtro indices o formato de retorno

# Linha 1
# indice 1 = valor 2
# indice 5 = valor 6
# Linha 2
# indice 10 = valor 11
# indice 12 = valor 13

array_3d.take(indices=[[1, 5], [10, 12]])

array([[ 2,  6],
       [11, 13]])

### `put`

Muda o valor contído no índice solicitado

In [ ]:
array_3d.put(indices=[1, 5, 10, 12], values=[52, 53, 54, 55])
array_3d

array([[[ 1, 52,  3],
        [ 4,  5, 53],
        [ 7,  8,  9],
        [10, 54, 12]],

       [[55, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

### `repeat`

Repete cada valor do array após ele mesmo

In [ ]:
# recriando o array original
array_3d = np.arange(1, 25).reshape(2, 4, 3)
array_3d

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_3d.shape

(2, 4, 3)

In [ ]:
# repete cada número 2 vezes
# por padrão o resultado é linearizado
array_3d.repeat(repeats=2)

array([ 1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,  8,  9,
        9, 10, 10, 11, 11, 12, 12, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17,
       18, 18, 19, 19, 20, 20, 21, 21, 22, 22, 23, 23, 24, 24])

In [ ]:
# A repetição aconteceu ao longo do eixo 0 (profundidade)
# portanto as matrizes são duplicadas
array_3d_repeat_2_axis_0 = array_3d.repeat(repeats=2, axis=0)
array_3d_repeat_2_axis_0

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

In [ ]:
array_3d_repeat_2_axis_0.shape

(4, 4, 3)

In [ ]:
# A repetição aconteceu ao longo do eixo 1 (linha)
# portanto as linhas são duplicadas
array_3d_repeat_2_axis_1 = array_3d.repeat(repeats=2, axis=1)
array_3d_repeat_2_axis_1

array([[[ 1,  2,  3],
        [ 1,  2,  3],
        [ 4,  5,  6],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [ 7,  8,  9],
        [10, 11, 12],
        [10, 11, 12]],

       [[13, 14, 15],
        [13, 14, 15],
        [16, 17, 18],
        [16, 17, 18],
        [19, 20, 21],
        [19, 20, 21],
        [22, 23, 24],
        [22, 23, 24]]])

In [ ]:
# o primeiro elemento repetirá 1 vez
# o segundo elemento repetirá 2 vezes
repeats = [1, 2]

# ao longo do eixo 0 (profundidade)
axis = 0

array_3d.repeat(repeats=repeats, axis=axis)

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]],

       [[13, 14, 15],
        [16, 17, 18],
        [19, 20, 21],
        [22, 23, 24]]])

### `sort`

O comando sort é utilizado para a ordenação de arrays

In [ ]:
# criando um array de exemplo
array_2d = np.array([[10, 3, 7], [8, 2, 9]])
array_2d

array([[10,  3,  7],
       [ 8,  2,  9]])

In [ ]:
array_2d.shape

(2, 3)

In [ ]:
# axis = 0, indica que queremos fazer as operaçẽos baseadas nas linhas
array_2d.sort(axis=0)
array_2d

array([[ 8,  2,  7],
       [10,  3,  9]])

In [ ]:
# recriando um array de exemplo
array_2d = np.array([[10, 3, 7], [8, 2, 9]])
array_2d

array([[10,  3,  7],
       [ 8,  2,  9]])

In [ ]:
# axis = 1, indica que queremos fazer as operaçẽos baseadas nas colunas
array_2d.sort(axis=1)
array_2d

array([[ 3,  7, 10],
       [ 2,  8,  9]])

### `argsort`

Retorna os índices que resultariam em um vetor ordenado

In [ ]:
# recriando um array de exemplo
array_2d = np.array([[10, 3, 7], [8, 2, 9]])
array_2d

array([[10,  3,  7],
       [ 8,  2,  9]])

In [ ]:
# retornando os indices que resultariam em um vetor ordenado
# ao longo das linhas
indices = array_2d.argsort(axis=0)
indices

array([[1, 1, 0],
       [0, 0, 1]])

In [ ]:
# retornando os indices que resultariam em um vetor ordenado
# ao longo das colunas
indices = array_2d.argsort(axis=1)
indices

array([[1, 2, 0],
       [1, 0, 2]])

In [ ]:
# utilizando um array de 1 dimensão para entender melhor

# criando um array 1d
array_1d = np.array([10, 3, 7, 8, 2, 9])
indices = array_1d.argsort()
indices

array([4, 1, 2, 3, 5, 0])

In [ ]:
# aplicando os indices para ordenar o array
array_1d[indices]

array([ 2,  3,  7,  8,  9, 10])

### `searchsort`

Searchsort procura em quais indices de um array ordenado você deve inserir os elementos de um segundo array para manter ordenado.

In [ ]:
# utilizando um array de 1 dimensão para entender melhor

# criando um array 1d já ordenda
# indices:            0  1  2  3  4  5
array_1d = np.array([ 2, 3, 7, 8, 9, 10])

indices = array_1d.searchsorted([1, 4, 11]) # [1, 4, 11] é o segundo array
indices

array([0, 2, 6])

In [ ]:
# aplicando os indices para ordenar o array
array_1d[indices]

### `searchsort`

Retorna a diagonal do array

In [ ]:
# criando um array 2d
array_2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
array_2d

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
array_2d.diagonal()

array([1, 5, 9])

Documentação dos métodos de seleção e manipulação de itens: https://numpy.org/doc/stable/reference/arrays.ndarray.html#item-selection-and-manipulation

# Exercícios

Vamos começar instanciando o mesmo array do módulo anterior, que representa o desempenho de 5 atletas em 4 voltas de uma pista de corrida, em 3 dias diferentes.

In [1]:
import numpy as np

In [3]:
rng = np.random.default_rng(seed=1)

In [4]:
array = rng.normal(loc=50, scale=3, size=60).reshape(3, 5, 4).round(3)
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [5]:
array.shape

(3, 5, 4)

👆 Observe pelo `shape` que são 3 dias, 5 atletas e 4 voltas.

## Métodos de conversão

### `item`

Utilize o método `item` para recuperar o desempenho do segundo atleta no terceiro dia, na primeira volta.

In [6]:
array.item((2, 1, 0))

49.705

### `itemset`

Suponha que este valor tenha sido registrado errado, e que o valor correto seja 49.709. Faça a correção utilizando `itemset`, e exiba o array corrigido.

In [7]:
array.itemset((2, 1, 0), 49.709)
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.709, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Utilize `item` de novo para ter certeza de que o valor foi alterado.

In [8]:
array.item((2, 1, 0))

49.709

### `view`

Gere uma view deste array, atribuindo o resultado a um novo array `array_view`.

In [9]:
array_view = array.view()
array_view

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.709, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Utilize `itemset` para mudar o valor corrigido anteriormente (índice (2, 1, 0)) para seu valor original (49.705), apenas em `array_view`.

In [10]:
array_view.itemset((2, 1, 0), 49.705)
array_view

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Agora, utilize `item` para verificar o valor no índice (2, 1, 0) em `array` e `array_view`.

In [11]:
array.item((2, 1, 0))

49.705

In [12]:
array_view.item((2, 1, 0))

49.705

💡 Observe que, apesar de você ter feito a modificação em `array_view`, `array` também foi modificado. Isso acontece porque, sendo `array_view` uma **view** de `array`, seus dados são compartilhados.

### `copy`

Agora repita o exercício anterior, mas criando uma cópia de `array`, armazenada em `array_copy`.

In [13]:
array_copy = array.copy()
array_copy

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Mude de novo o valor em (2, 1, 0) para 49.709, somente na cópia.

In [14]:
array_copy.itemset((2, 1, 0), 49.709)
array_copy

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.709, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Agora, utilize `item` para verificar o valor no índice (2, 1, 0) em `array` e `array_copy`.

In [17]:
array.item((2, 1, 0)) # não foi modificado

49.705

In [16]:
array_copy.item((2, 1, 0))

49.709

💡 Uma **cópia** de um array mantém dados independentes do array original.

## Métodos de manipulação de shape

### `reshape`

Suponha que você precise registrar os resultados do desempenho dos atletas em uma planilha, que armazena dados em apenas 2 dimensões. A orientação que você recebeu é que cada linha da planilha deve ser correspondente a um dia, totalizando 3 linhas, e as colunas vão conter as 4 voltas do primeiro atleta, depois as 4 voltas do segundo atleta, e assim por diante, totalizando 5 atletas x 4 voltas = 20 colunas.

Exiba uma versão do array nesta configuração.

In [21]:
array.reshape(3, 20)
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

### `transpose`

Lembre-se que as dimensões do array representam os dias, os atletas e as voltas, nesta ordem.

Utilize `transpose` para representar o array com os dados dos atletas, depois as voltas, depois os dias. Armazene o resultado em um novo `array_transpose`, exiba o array e, depois seu `shape`.

In [22]:
array_transpose = array.transpose((1, 2, 0))
array_transpose

array([[[51.037, 50.024, 46.318],
        [52.465, 49.173, 47.95 ],
        [50.991, 53.882, 49.784],
        [46.091, 53.02 , 47.166]],

       [[52.716, 41.867, 49.705],
        [51.339, 44.333, 50.286],
        [48.389, 49.476, 50.107],
        [51.743, 48.733, 48.481]],

       [[51.094, 50.641, 51.781],
        [50.882, 50.652, 52.674],
        [50.085, 56.354, 50.963],
        [51.64 , 46.664, 47.545]],

       [[47.791, 48.867, 52.195],
        [49.511, 56.128, 48.496],
        [48.554, 51.94 , 52.637],
        [51.797, 51.989, 46.785]],

       [[50.119, 48.458, 52.743],
        [49.123, 45.056, 49.94 ],
        [47.654, 50.502, 46.254],
        [49.228, 50.327, 49.058]]])

In [23]:
array_transpose.shape

(5, 4, 3)

👆 5 atletas, 4 voltas, 3 dias.

💡 Ou seja, a primeira linha da visualização:
> ```
> [51.037, 50.024, 46.318]
> ```

representa o primeiro atleta, a primeira volta, e os 3 dias.

### `swapaxes`

A partir de `array_transpose`, agora utilize `swapaxes` para mudar os eixos de forma a representar os atletas, depois os dias, depois as voltas. Armazene o resultado em `array_swapaxes`, exiba o array e seu shape.

In [24]:
array_swapaxes = array_transpose.swapaxes(2, 1)
array_swapaxes

array([[[51.037, 52.465, 50.991, 46.091],
        [50.024, 49.173, 53.882, 53.02 ],
        [46.318, 47.95 , 49.784, 47.166]],

       [[52.716, 51.339, 48.389, 51.743],
        [41.867, 44.333, 49.476, 48.733],
        [49.705, 50.286, 50.107, 48.481]],

       [[51.094, 50.882, 50.085, 51.64 ],
        [50.641, 50.652, 56.354, 46.664],
        [51.781, 52.674, 50.963, 47.545]],

       [[47.791, 49.511, 48.554, 51.797],
        [48.867, 56.128, 51.94 , 51.989],
        [52.195, 48.496, 52.637, 46.785]],

       [[50.119, 49.123, 47.654, 49.228],
        [48.458, 45.056, 50.502, 50.327],
        [52.743, 49.94 , 46.254, 49.058]]])

In [25]:
array_swapaxes.shape

(5, 3, 4)

👆 5 atletas, 3 dias, 4 voltas.

💡 Ou seja, a primeira linha da visualização:
> ```
> [51.037, 52.465, 50.991, 46.091]
> ```

representa o primeiro atleta, o primeiro dia, as 4 voltas.

### `flatten` / `ravel`

Agora suponha que você tenha que registrar os resultados em um sistema que foi programado para receber dados de forma linear, um valor por vez.

Linearize o array utilizando `flatten` e `ravel`.

In [28]:
array.flatten() # copy

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354, 46.664,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
       46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

In [29]:
array.ravel() # view

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354, 46.664,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
       46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

💡 Lembre-se que `flatten` retorna uma **cópia** do array, enquanto que `ravel` retorna uma **view**, que compartilha os dados do `array`, mas mantém seu próprio `shape`.

### `squeeze`

Nós podemos separar os resultados do primeiro dia em um novo array `array_first_day`, com a seguinte notação:

In [30]:
array_first_day = array[0:1]
array_first_day

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]]])

In [31]:
array_first_day.ndim, array_first_day.shape

(3, (1, 5, 4))

Observe que o número de dimensões (`ndim`) de `array_first_day` continua sendo igual a 3, ou seja, o mesmo de `array`. Com `shape`, nós observamos que o eixo correspondente aos dias possui um único elemento, justamente o primeiro dia selecionado.

Utilize `squeeze` para simplificar este array, removendo o eixo dos dias, já que ele contém somente um único dia. Salve o resultado em `array_squeeze`, exiba o array e, depois, seu `shape`.

In [33]:
array_squeeze = array_first_day.squeeze()
array_squeeze

array([[51.037, 52.465, 50.991, 46.091],
       [52.716, 51.339, 48.389, 51.743],
       [51.094, 50.882, 50.085, 51.64 ],
       [47.791, 49.511, 48.554, 51.797],
       [50.119, 49.123, 47.654, 49.228]])

👆 Veja que tem um par de colchetes a menos, o que indica a redução de 1 eixo.

In [35]:
array_squeeze.shape

(5, 4)

👆 O eixo correspondente ao dia, de tamanho 1, foi removido.

## Métodos de seleção e manipulação de itens

### `take`

Agora considere que os dados em `array` se referem ao mesmo atleta. O primeiro eixo continua representando 3 dias diferentes, o último representa as 4 voltas de uma prova, mas o segundo eixo representa 5 provas completas do mesmo atleta. Ou seja, `array` agora contém dados consecutivos do mesmo atleta.

Suponha que, assim que terminou as provas, o atleta se queixou de dor no joelho nas voltas de número 12, 25, 31 e 40. Utilizando `take`, recupere o desempenho do atleta nessas voltas.

In [36]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [37]:
array.take(indices=[12, 24, 31, 40])

array([47.791, 41.867, 46.664, 46.318])

Mas o atleta foi mais específico e disse que, nas voltas 12 e 31, a dor foi no joelho direito, e nas voltas 25 e 40, a dor foi no joelho esquerdo. Utilizando `take`, recupere essas voltas novamente, mas de forma a obter os resultados como um array onde a primeira linha se refira ao joelho direito, e a segunda ao joelho esquerdo.

In [38]:
array.take(indices=[[12, 24], [31, 40]])

array([[47.791, 41.867],
       [46.664, 46.318]])

### `put`

Com base no que o atleta reportou, seu treinador decidiu desconsiderar o desempenho do atleta nas voltas onde ele se queixou de dor. Utilize `put` para mudar estes valores para a variável `np.nan`, que representa valores nulos em um array do NumPy. Depois, exiba o array modificado.

In [39]:
array.put(indices=[12, 24, 31, 40], values=np.nan)
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

### `repeat`

Repita os valores do array para que eles sejam duplicados (`repeat = 2`), ao longo do eixo 0, 1 e 2, separadamente.

Depois, exiba o array para inspecionar onde ocorreu a repetição. Exiba também o `shape` do array para observar a alteração feita.

In [40]:
array_repeat_axis_0 = array.repeat(repeats=2, axis=0)
array_repeat_axis_0

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 5

In [41]:
array_repeat_axis_0.shape

(6, 5, 4)

👆 Veja que os dados correspondentes aos dias foram duplicados.

In [42]:
array_repeat_axis_1 = array.repeat(repeats=2, axis=1)
array_repeat_axis_1

array([[[51.037, 52.465, 50.991, 46.091],
        [51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [49.705, 50.286, 50.10

In [43]:
array_repeat_axis_1.shape

(3, 10, 4)

👆 Veja que os dados correspondentes às provas foram duplicados.

In [44]:
array_repeat_axis_2 = array.repeat(repeats=2, axis=2)
array_repeat_axis_2

array([[[51.037, 51.037, 52.465, 52.465, 50.991, 50.991, 46.091, 46.091],
        [52.716, 52.716, 51.339, 51.339, 48.389, 48.389, 51.743, 51.743],
        [51.094, 51.094, 50.882, 50.882, 50.085, 50.085, 51.64 , 51.64 ],
        [   nan,    nan, 49.511, 49.511, 48.554, 48.554, 51.797, 51.797],
        [50.119, 50.119, 49.123, 49.123, 47.654, 47.654, 49.228, 49.228]],

       [[50.024, 50.024, 49.173, 49.173, 53.882, 53.882, 53.02 , 53.02 ],
        [   nan,    nan, 44.333, 44.333, 49.476, 49.476, 48.733, 48.733],
        [50.641, 50.641, 50.652, 50.652, 56.354, 56.354,    nan,    nan],
        [48.867, 48.867, 56.128, 56.128, 51.94 , 51.94 , 51.989, 51.989],
        [48.458, 48.458, 45.056, 45.056, 50.502, 50.502, 50.327, 50.327]],

       [[   nan,    nan, 47.95 , 47.95 , 49.784, 49.784, 47.166, 47.166],
        [49.705, 49.705, 50.286, 50.286, 50.107, 50.107, 48.481, 48.481],
        [51.781, 51.781, 52.674, 52.674, 50.963, 50.963, 47.545, 47.545],
        [52.195, 52.195, 48.496, 4

In [45]:
array_repeat_axis_2.shape

(3, 5, 8)

👆 Veja que os dados correspondentes às voltas foram duplicados.

### `sort`

Primeiro, crie uma cópia de `array`, já que `sort` altera o array *inplace*.

Depois, utilize `sort` para ordenar o array ao longo do eixo 0. Exiba o array ordenado para observar o que aconteceu.

In [46]:
array_copy = array.copy()
array_copy

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [47]:
array_copy.sort(axis=0)
array_copy

array([[[50.024, 47.95 , 49.784, 46.091],
        [49.705, 44.333, 48.389, 48.481],
        [50.641, 50.652, 50.085, 47.545],
        [48.867, 48.496, 48.554, 46.785],
        [48.458, 45.056, 46.254, 49.058]],

       [[51.037, 49.173, 50.991, 47.166],
        [52.716, 50.286, 49.476, 48.733],
        [51.094, 50.882, 50.963, 51.64 ],
        [52.195, 49.511, 51.94 , 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[   nan, 52.465, 53.882, 53.02 ],
        [   nan, 51.339, 50.107, 51.743],
        [51.781, 52.674, 56.354,    nan],
        [   nan, 56.128, 52.637, 51.989],
        [52.743, 49.94 , 50.502, 50.327]]])

👆 Observe que, ao longo dos dias, os valores estão ordenados de forma crescente: `50.024 -> 51.037 -> nan`, `47.95 -> 49.173 -> 52.465`...

Também observe que o NumPy coloca `nan` no **final** da ordem; todos os valores `nan` estão no terceiro elemento do eixo 0.

Agora repita o procedimento, copiando o array e depois ordenando ao longo do eixo 1.

In [48]:
array_copy = array.copy()
array_copy

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [49]:
array_copy.sort(axis=1)
array_copy

array([[[50.119, 49.123, 47.654, 46.091],
        [51.037, 49.511, 48.389, 49.228],
        [51.094, 50.882, 48.554, 51.64 ],
        [52.716, 51.339, 50.085, 51.743],
        [   nan, 52.465, 50.991, 51.797]],

       [[48.458, 44.333, 49.476, 48.733],
        [48.867, 45.056, 50.502, 50.327],
        [50.024, 49.173, 51.94 , 51.989],
        [50.641, 50.652, 53.882, 53.02 ],
        [   nan, 56.128, 56.354,    nan]],

       [[49.705, 47.95 , 46.254, 46.785],
        [51.781, 48.496, 49.784, 47.166],
        [52.195, 49.94 , 50.107, 47.545],
        [52.743, 50.286, 50.963, 48.481],
        [   nan, 52.674, 52.637, 49.058]]])

👆 Agora, a ordenação ocorre ao longo das provas (que na visualização correspondem às colunas): `50.119 -> 51.037 -> 51.094 -> 52.716 -> nan`...

Desta vez, os valores `nan` estão no último elemento do eixo 1 (última linha de cada array que representa os dias).

Repita mais uma vez, copiando o array e depois ordenando ao longo do eixo 2.

In [50]:
array_copy = array.copy()
array_copy

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [   nan, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [52]:
array_copy.sort(axis=2)
array_copy

array([[[46.091, 50.991, 51.037, 52.465],
        [48.389, 51.339, 51.743, 52.716],
        [50.085, 50.882, 51.094, 51.64 ],
        [48.554, 49.511, 51.797,    nan],
        [47.654, 49.123, 49.228, 50.119]],

       [[49.173, 50.024, 53.02 , 53.882],
        [44.333, 48.733, 49.476,    nan],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 51.94 , 51.989, 56.128],
        [45.056, 48.458, 50.327, 50.502]],

       [[47.166, 47.95 , 49.784,    nan],
        [48.481, 49.705, 50.107, 50.286],
        [47.545, 50.963, 51.781, 52.674],
        [46.785, 48.496, 52.195, 52.637],
        [46.254, 49.058, 49.94 , 52.743]]])

👆 A ordenação ocorre ao longo das voltas (que na visualização correspondem às linhas): `46.091 -> 50.991 -> 51.037 -> 52.465`...

Desta vez, os valores `nan` estão no último elemento do eixo 2 (última coluna  de cada array que representa os dias).

### `argsort`

Repita o exercício com `sort`, mas agora, ao invés de ordenar os valores *inplace*, recupere e exiba os índices necessários para ordenar o array. Isto é feito com o método `argsort`.

Desta vez, como não há modificação do array, você não precisa da cópia.

In [53]:
indices_axis_0 = array.argsort(axis=0)
indices_axis_0

array([[[1, 2, 2, 0],
        [2, 1, 0, 2],
        [1, 1, 0, 2],
        [1, 2, 0, 2],
        [1, 1, 2, 2]],

       [[0, 1, 0, 2],
        [0, 2, 1, 1],
        [0, 0, 2, 0],
        [2, 0, 1, 0],
        [0, 0, 0, 0]],

       [[2, 0, 1, 1],
        [1, 0, 2, 0],
        [2, 2, 1, 1],
        [0, 1, 2, 1],
        [2, 2, 1, 1]]])

👆 Observe que o valor máximo neste array que representa os índices é 2, porque o eixo 0 só contém 3 elementos (3 dias).

In [54]:
indices_axis_1 = array.argsort(axis=1)
indices_axis_1

array([[[4, 4, 4, 0],
        [0, 3, 1, 4],
        [2, 2, 3, 2],
        [1, 1, 2, 1],
        [3, 0, 0, 3]],

       [[4, 1, 1, 1],
        [3, 4, 4, 4],
        [0, 0, 3, 3],
        [2, 2, 0, 0],
        [1, 3, 2, 2]],

       [[1, 0, 4, 3],
        [2, 3, 0, 0],
        [3, 4, 1, 2],
        [4, 1, 2, 1],
        [0, 2, 3, 4]]])

👆 Agora o valor máximo é 4, pois o eixo 1 tem 5 elementos (5 atletas).

In [55]:
indices_axis_2 = array.argsort(axis=2)
indices_axis_2

array([[[3, 2, 0, 1],
        [2, 1, 3, 0],
        [2, 1, 0, 3],
        [2, 1, 3, 0],
        [2, 1, 3, 0]],

       [[1, 0, 3, 2],
        [1, 3, 2, 0],
        [0, 1, 2, 3],
        [0, 2, 3, 1],
        [1, 0, 3, 2]],

       [[3, 1, 2, 0],
        [3, 0, 2, 1],
        [3, 2, 0, 1],
        [3, 1, 0, 2],
        [2, 3, 1, 0]]])

👆 O valor máximo é 3, pois o eixo 2 tem 4 elementos (4 voltas).

### `searchsorted`

Este método só funciona para arrays 1D que já estejam ordenados.

Então, primeiro linearize `array` com `flatten`, depois ordene com `sort` (lembre-se que o método `sort` altera o array *inplace*, ou seja, ele não retorna nada, então você não precisa armazenar o resultado em uma nova variável), e utilize indexação para recuperar os primeiros 15 elementos.

In [56]:
array_flatten = array.flatten()
array_flatten

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 ,    nan, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
          nan, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354,    nan,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
          nan, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

In [57]:
array_flatten.sort()
array_flatten

array([44.333, 45.056, 46.091, 46.254, 46.785, 47.166, 47.545, 47.654,
       47.95 , 48.389, 48.458, 48.481, 48.496, 48.554, 48.733, 48.867,
       49.058, 49.123, 49.173, 49.228, 49.476, 49.511, 49.705, 49.784,
       49.94 , 50.024, 50.085, 50.107, 50.119, 50.286, 50.327, 50.502,
       50.641, 50.652, 50.882, 50.963, 50.991, 51.037, 51.094, 51.339,
       51.64 , 51.743, 51.781, 51.797, 51.94 , 51.989, 52.195, 52.465,
       52.637, 52.674, 52.716, 52.743, 53.02 , 53.882, 56.128, 56.354,
          nan,    nan,    nan,    nan])

In [58]:
array_filtered = array_flatten[:15]
array_filtered

array([44.333, 45.056, 46.091, 46.254, 46.785, 47.166, 47.545, 47.654,
       47.95 , 48.389, 48.458, 48.481, 48.496, 48.554, 48.733])

In [59]:
indices = array_filtered.searchsorted([40.50, 42.357, 47.025, 48.952])
indices

array([ 0,  0,  5, 15])

👆 Inspecione o array e os índices encontrados para confirmar a resposta.

### `searchsorted`

O entendimento de diagonais é facilitado trabalhando com arrays 2D.

Por isso, primeiro obtenha com indexação uma view do array correspondente ao primeiro dia.

In [61]:
array_2d = array[0]
array_2d

array([[51.037, 52.465, 50.991, 46.091],
       [52.716, 51.339, 48.389, 51.743],
       [51.094, 50.882, 50.085, 51.64 ],
       [   nan, 49.511, 48.554, 51.797],
       [50.119, 49.123, 47.654, 49.228]])

In [62]:
array_2d.shape

(5, 4)

Agora, recupere e exiba a diagonal deste array.

In [64]:
array_2d.diagonal()

array([51.037, 51.339, 50.085, 51.797])

👆 Não é difícil encontrar visualmente esta diagonal em `array_2D`.

Observe que, como `array_2D` é **retangular** (como demonstra seu `shape`), nenhum valor da última linha é retornado na diagonal, pois ele tem mais linhas que colunas.

#Métodos para Cálculo

### `max`

Extrai o maior valor do array

In [65]:
# criando um array 2d como exemplo
array_2d = np.array([[ 4,  2, 12],
                     [10,  5,  6],
                     [ 1, 11,  9],
                     [ 7,  8,  3]])

array_2d

array([[ 4,  2, 12],
       [10,  5,  6],
       [ 1, 11,  9],
       [ 7,  8,  3]])

In [75]:
# valor maximo de todo o array
array_2d.max()

12

In [69]:
# axis = 0 (valor máximo de cada coluna)
array_2d.max(axis=0)

array([10, 11, 12])

In [70]:
# axis = 1 (valor máximo de cada linha)
array_2d.max(axis=1)

array([12, 10, 11,  8])

In [72]:
# retorna o valor máximo de cada linha mantendo as dimensões do array
array_2d_axis_1_kd = array_2d.max(axis=1, keepdims=True)
array_2d_axis_1_kd

array([[12],
       [10],
       [11],
       [ 8]])

In [74]:
array_2d_axis_1_kd.shape, array_2d.shape

((4, 1), (4, 3))

### `argmax`

Retorna os índices de onde estão os maiores valores

In [76]:
array_2d

array([[ 4,  2, 12],
       [10,  5,  6],
       [ 1, 11,  9],
       [ 7,  8,  3]])

In [83]:
# retorna o indíce do maior valor de todo o array
array_2d.argmax()

2

In [79]:
# retorna o índice da linha que tem o maior elemento por coluna
array_2d.argmax(axis=0)

array([1, 2, 0])

In [81]:
# retorna o índice da coluna que tem o maior elemento por linha
array_2d.argmax(axis=1)

array([2, 0, 1, 1])

### `min`

In [84]:
array_2d

array([[ 4,  2, 12],
       [10,  5,  6],
       [ 1, 11,  9],
       [ 7,  8,  3]])

In [85]:
# retorna o valor mínimo do array
array_2d.min()

1

In [86]:
# retorna o valor mínimo do array por coluna
array_2d.min(axis=0)

array([1, 2, 3])

In [88]:
# retorna o valor mínimo do array por linha
array_2d.min(axis=1)

array([2, 5, 1, 3])

### `argmin`

In [91]:
array_2d

array([[ 4,  2, 12],
       [10,  5,  6],
       [ 1, 11,  9],
       [ 7,  8,  3]])

In [89]:
# retorna o indice com o valor mínimo do array
array_2d.argmin()

6

In [90]:
# retorna o índice da linha que tem o menor elemento por coluna
array_2d.argmin(axis=0)

array([2, 0, 3])

In [92]:
# retorna o índice da coluna que tem o menor elemento por linha
array_2d.argmin(axis=1)

array([1, 1, 0, 2])

### `ptp (peak to peak)`

Retorna a subtração do valor máximo pelo valor mínimo.

In [93]:
array_2d

array([[ 4,  2, 12],
       [10,  5,  6],
       [ 1, 11,  9],
       [ 7,  8,  3]])

In [95]:
# valor máximo menos o valor mínimo de todo o array
array_2d.max(), array_2d.min(), array_2d.ptp()

(12, 1, 11)

In [96]:
# valor máximo menos o valor mínimo por coluna
array_2d.max(axis = 0), array_2d.min(axis = 0), array_2d.ptp(axis=0)

(array([10, 11, 12]), array([1, 2, 3]), array([9, 9, 9]))

In [97]:
# valor máximo menos o valor mínimo por linha
array_2d.max(axis = 1), array_2d.min(axis = 1), array_2d.ptp(axis=1)

(array([12, 10, 11,  8]), array([2, 5, 1, 3]), array([10,  5, 10,  5]))

### `clip`

Converte os valores menores que o mímimo para um valor mínimo específicado e os valores maiores que máximos para o máximo especificado.

In [99]:
array_2d = np.arange(1, 13).reshape(4, 3)
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [101]:
# converte os valores menores que 3 para 3
# converte os valores maiores que 10 para 10

array_2d.clip(min=3, max=10)

array([[ 3,  3,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 10, 10]])

### `round`

Arrendonda os valores para o numero de casas decimais solicitados.

In [102]:
array_2d = np.linspace(0, 1, 12).reshape(4, 3)
array_2d

array([[0.        , 0.09090909, 0.18181818],
       [0.27272727, 0.36363636, 0.45454545],
       [0.54545455, 0.63636364, 0.72727273],
       [0.81818182, 0.90909091, 1.        ]])

In [103]:
# arredonda para duas casas decimais
array_2d.round(decimals=2)

array([[0.  , 0.09, 0.18],
       [0.27, 0.36, 0.45],
       [0.55, 0.64, 0.73],
       [0.82, 0.91, 1.  ]])

### `sum`

Soma dos elementos do vetor ou do eixo solicitado

In [104]:
array_2d = np.arange(1, 13).reshape(4, 3)
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [105]:
# Soma todos os itens do array
array_2d.sum()

78

In [106]:
# soma todos os itens por coluna
array_2d.sum(axis = 0)

array([22, 26, 30])

In [108]:
# soma todos os itens por linha
array_2d.sum(axis = 1)

array([ 6, 15, 24, 33])

### `cumsum`

Faz a soma acumulada de todos os itens de um array ou de um eixo especificado.

In [109]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [110]:
# soma acumulada de todos os itens do array
array_2d.cumsum()

array([ 1,  3,  6, 10, 15, 21, 28, 36, 45, 55, 66, 78])

In [111]:
# soma acumulada de todos os itens do array por coluna
array_2d.cumsum(axis=0)

array([[ 1,  2,  3],
       [ 5,  7,  9],
       [12, 15, 18],
       [22, 26, 30]])

In [112]:
# soma acumulada de todos os itens do array por linha
array_2d.cumsum(axis=1)

array([[ 1,  3,  6],
       [ 4,  9, 15],
       [ 7, 15, 24],
       [10, 21, 33]])

### `mean`

Calcula a média dos elementos de uma array ou de um eixo especificado.

In [114]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [113]:
# calcula a média dos valores de um array
array_2d.mean()

6.5

In [116]:
# calcula a média dos valores por coluna
array_2d.mean(axis=0)

array([5.5, 6.5, 7.5])

In [117]:
# calcula a média dos valores por linha
array_2d.mean(axis=1)

array([ 2.,  5.,  8., 11.])

### `var`

Calcula a variância dos elementos de uma array ou de um eixo especificado.

In [118]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [120]:
# calcula a variância dos valores de um array
array_2d.var()

11.916666666666666

In [121]:
# calcula a variância dos valores de um array por coluna
array_2d.mean(axis=0)

array([5.5, 6.5, 7.5])

In [122]:
# calcula a variância dos valores de um array por linha
array_2d.mean(axis=1)

array([ 2.,  5.,  8., 11.])

### `std`

Calcula o desvio padrão dos elementos de uma array ou de um eixo especificado.

In [124]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [123]:
# calcula o desvio padrão dos valores de um array
array_2d.std()

3.452052529534663

In [125]:
# calcula o desvio padrão dos valores de um array por coluna
array_2d.std(axis=0)

array([3.35410197, 3.35410197, 3.35410197])

In [126]:
# calcula o desvio padrão dos valores de um array por linha
array_2d.std(axis=1)

array([0.81649658, 0.81649658, 0.81649658, 0.81649658])

### `prod`

Retorna o produto de todos os elementos de um vetor ou de um eixo especificado.

In [127]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [128]:
# calcula o produto dos valores de um array
array_2d.prod()

479001600

In [130]:
# calcula o produto dos valores de um array por coluna
array_2d.prod(axis=0)

array([ 280,  880, 1944])

In [131]:
# calcula o produto dos valores de um array por linha
array_2d.prod(axis=1)

array([   6,  120,  504, 1320])

### `cumprod`

Retorna o produto acumulado de todos os elementos de um vetor ou de um eixo especificado.

In [132]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [133]:
# calcula o produto acumulado dos valores de um array
array_2d.cumprod()

array([        1,         2,         6,        24,       120,       720,
            5040,     40320,    362880,   3628800,  39916800, 479001600])

In [134]:
# calcula o produto acumulado dos valores de um array por coluna
array_2d.cumprod(axis=0)

array([[   1,    2,    3],
       [   4,   10,   18],
       [  28,   80,  162],
       [ 280,  880, 1944]])

In [136]:
# calcula o produto acumulado dos valores de um array por linha
array_2d.cumprod(axis=1)

array([[   1,    2,    6],
       [   4,   20,  120],
       [   7,   56,  504],
       [  10,  110, 1320]])

### `all`

Indica que todos os elementos do vetor serão verificados.

In [137]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [138]:
# Verificação se os valores são maiores que 0
array_2d_gt_0 = array_2d > 0
array_2d_gt_0

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [139]:
# Verificação se os valores são maiores que 0 em todo resultado
array_2d_gt_0.all()

True

In [140]:
# Verificação se os valores são maiores que 0 por coluna
array_2d_gt_0.all(axis=0)

array([ True,  True,  True])

In [141]:
# Verificação se os valores são maiores que 0 por linha
array_2d_gt_0.all(axis=1)

array([ True,  True,  True,  True])

In [142]:
# Verificação se os valores são maiores que 5
array_2d_gt_5 = array_2d > 5
array_2d_gt_5

array([[False, False, False],
       [False, False,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [145]:
# Verificação se os valores são maiores que 5 em todo resultado
array_2d_gt_5.all()

False

In [147]:
# Verificação se os valores são maiores que 5 por coluna
array_2d_gt_5.all(axis=0)

array([False, False, False])

In [148]:
# Verificação se os valores são maiores que 5 por linha
array_2d_gt_5.all(axis=1)

array([False, False,  True,  True])

### `any`

Testa se há ao mesmos um valor True

In [150]:
array_2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [151]:
# Verificação se os valores são maiores que 5
array_2d_gt_5 = array_2d >5
array_2d_gt_5

array([[False, False, False],
       [False, False,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [152]:
# verifica se há ao menos um True em todo o array
array_2d_gt_5.any()

True

In [153]:
# verifica se há ao menos um True por coluna
array_2d_gt_5.any(axis=0)

array([ True,  True,  True])

In [154]:
# verifica se há ao menos um True por linha
array_2d_gt_5.any(axis=1)

array([False,  True,  True,  True])

Documentação dos métodos para cálculo: https://numpy.org/doc/stable/reference/arrays.ndarray.html#calculation